src/app/CommandSender.h

```diff
@@ -36,15 +38,71 @@
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received and it is Success, aData will be nullptr.
+         *  - When command specific response is received, aData will be a valid TLVReader initialized with a struct of command
+         * specify response.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction continue to exist after
+         * this call completes.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aData: The command data, will be nullptr if the server returns a StatusElement.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath, TLV::TLVReader * aData) {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         * - CHIP_ERROR_IM: The CommandSender received a legal response from server, but the protocol code indicates a error status
+         * presended in aInteractionModelStatus.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction,  continue to exist after
+         * this call completes.
+         * @param[in] aInteractionModelStatus: The interaction model protocol status code, will be Failure if the CommandSender
+         * failed to recieve the response (aError != CHIP_NO_ERROR). This argument will never be Success when this function is
+         * called.
+         * @param[in] aProtocolError: The error occurred during the invoke command interaction, will be CHIP_NO_ERROR when the
+         * CommandSender received a valid response from server.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, Protocols::InteractionModel::Status aInteractionModelStatus,
+                             CHIP_ERROR aProtocolError)
+        {}
+
+        /**
+         * OnFinal will be called when CommandSender has finished all the works, and it is safe to release the
+         * allocated CommandSender object if needed.
+         *
+         * This function is marked as must be implemented to make application aware of releasing the object.
+         *
+         * After the command being sent, this function will always be called exactly once for a single command sender.
+         *
+         * @param[in] apCommandSender: The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnFinal(CommandSender * apCommandSender) = 0;
+    };
+
+    CommandSender(Callback * apCallback) : mpCallback(apCallback){};
+
+    ~CommandSender() { AbortExistingExchangeContext(); }
```

Comment:
The actual exchange context is 'owned' by the base class, so it feels more appropriate to abort it in `~Command()`.